Renaming commands

ellisonbg opened the pull request:
This PR does a review of the action names in the notebook to address #532.

We sat down with @sccolbert @jasongrout and @SylvainCorlay last week and spent a day going through all the action names, and seeing what patterns and inconsistencies there were. Here are the broad rules we are using for these names:

* First pick a noun and a verb for the action. For example, if the action is "restart kernel," the verb is "restart" and the noun is "kernel".
* Omit terms like "selected" and "active" by default, so "delete-cell", rather than "delete-selected-cell". Only provide a scope like "-all-" if it is other than the default "selected" or "active" scope.
* If an action has a secondary action, separate the secondary action with "-and-", so "restart-kernel-and-clear-output".
* Don't ever use before/after as they have a temporal connotation that is confusing when used in a spatial context.
* Use above/below or previous/next to indicate spacial and sequential relationships.
* For dialogs, use a verb that indicates what the dialog will accomplish, such as "confirm-restart-kernel".

This PR also starts to use `:` for action groups (looking forward to phosphor) and `jupyter-notebook` as the default group.

ssanderson commented:
:+1: for having a consistent naming-style guide on these.

> - Use above/below or previous/next to indicate spacial and sequential relationships.

This sentence is a bit ambiguous to me.  Is the intent here equivalent to "Use above/below to indicate spacial relationships.  Use previous/next to indicate sequential relationships."?  Is sequential indended as a synonym for "temporally-sequential" here?

ellisonbg commented:
Yeah, when to use above/below or previous/next can be a bit sutble and that difficulty is reflected in the sentence structure.

Carreau commented:
> Omit terms like "selected" and "active" by default, so "delete-cell", rather than "delete-selected-cell". Only provide a scope like "-all-" if it is other than the default "selected" or "active" scope.

Why that ? Especially since we have multiple selection and anchor ? I fel a lot confused when searching for an action if there is not the "selected", or "all"

Carreau commented on docs/source/examples/Notebook/Custom Keyboard Shortcuts.ipynb:
```diff
@@ -114,7 +114,7 @@
     "%%javascript\n",
     "\n",
-    "Jupyter.keyboard_manager.command_shortcuts.add_shortcut('r', 'ipython.change-selected-cell-to-raw-cell');"
+    "Jupyter.keyboard_manager.command_shortcuts.add_shortcut('r', 'ipython.change-cell-to-raw');"
```
`ipython:` ?

ellisonbg replied:
Opps, good catch.

Carreau replied:
Actually `jupyter-notebook`

Carreau commented on notebook/static/notebook/js/actions.js:
```diff
@@ -131,7 +137,7 @@ define(function(require){
             }
         },
-        'go-to-command-mode': {
+        'enter-command-mode': {
```
For me (as a non english) "enter" read like the enter key.

ellisonbg replied:
Modes are like contexts, like rooms in a house. You enter rooms, you enter modes.

Carreau replied:
`switch` ? Many time I say "Enter command mode" to French speaker, they press Enter. I would like to avoid this confusion.

Carreau commented on notebook/static/notebook/js/actions.js:
```diff
@@ -211,21 +217,21 @@ define(function(require){
             }
         },
-        'paste-cell-before' : {
+        'paste-cell-above' : {
```
I'm -0.7 on above / below, i would prefer `previous/next` as this apply to non-vertical interface (like slides).

ellisonbg replied:
As mentioned above the differences between previous/next versus above/below are pretty subtle. However we are finding that we pick the ones that match the needed parts of speech. For example "paste cell next" is not a proper sentence where "past cell below" is. If you wanted to use "next" in a sentence you would have to say something like "paste cell into the next slot in the notebook", which is a much more cumbersome and verbose way of saying "paste cell below".

Carreau commented on notebook/static/notebook/js/actions.js:
```diff
@@ -423,6 +415,21 @@ define(function(require){
+        'toggle-toolbar':{
+            help: 'hide/show the toolbar',
+            handler : function(env){
+                $('div#maintoolbar').toggle();
```
Oh, please no, don't use jquery here. I've made effort for `env` to be accessible. This is the only thing in the file that use jQuery.

ellisonbg replied:
I didn't add the jQuery here, I think you did ;-) 

I was simply moving this block into another place in the file near other shortcuts like it.

Carreau replied:
Fair enough, I just didn't had time to remove it then :-P

jdfreder replied:
Might as well fix it while we are looking at it - @ellisonbg can you change it to `env.toolbar.element.toggle()` please?

ellisonbg replied:
I would but the toolbar isn't in the env :(

This action stuff is a huge mess - I am not going to dive into that and clean everything up...

jdfreder replied:
Hmm, that is misleading.  I would have thought `env` === `Jupyter` seeing that we access `env.notebook`.

Carreau commented on notebook/static/notebook/js/actions.js:
```diff
@@ -444,7 +451,7 @@ define(function(require){
                 return true;
             }
         },
-        'move-cursor-up-or-previous-cell':{
```
Why do you remove `or-previous-cell` ? This will definitively not be use through command palette, and only in API (even if it will be **shown** in CP), so why not be explicit about what it does ? Or we will get confused user where it will  get unsuspecting behavior like changing the cell selection.

ellisonbg replied:
I don't feel strongly about this - it was the consensus of the 4 of us that the `or-previous-cell` was logical enough that it could be assumed. Also, the `or-previous-cell` doesn't clearly express what is happening, which is more like "move cursor up in current or previous cell". More than willing to change this back though-even better would be to find a clear way of spelling it out.

Carreau replied:
I'm not too strongly attached to this one. I'm even wondering if we shouldn't hide some action from CP.

minrk replied:
add a key for actions that shouldn't be listed?

jdfreder replied:
> add a key for actions that shouldn't be listed?

This sounds like a good idea, I could imagine actions that are purely internal.

Carreau commented on notebook/static/notebook/js/actions.js:
```diff
@@ -463,7 +470,7 @@ define(function(require){
             }
         },
-        'move-cursor-down-or-next-cell':{
+        'move-cursor-down':{
```
idem

Carreau commented on notebook/static/notebook/js/actions.js:
```diff
@@ -545,13 +552,13 @@ define(function(require){
+        registry['jupyter-notebook:'+subkey] = {};
+        registry['jupyter-notebook:'+subkey].help = source[subkey].help||subkey.replace(/-/g,' ');
+        registry['jupyter-notebook:'+subkey].help_index = source[subkey].help_index;
+        registry['jupyter-notebook:'+subkey].icon = source[subkey].icon;
```
why not just "notebook" ?

Carreau replied:
Also I think the "builtin" or alike would be more appropriate. There is specific code in CP that do not show the category if in builtin or ipython NS.

ellisonbg replied:
We are looking ahead to the many separate plugins/components situation where there will be lots of jupyter-components (jupyter-terminal, jupyter-filebrowser) in addition to third party ones (bloomberg-this, continuum-that, etc.). We will need this level of namespacing in both KB shortcuts, CSS, etc.

Also, in a single page app, it is not clear that the "notebook" component is the "builtin" one - someone may use the Jupyter web app and never run a notebook.

Carreau replied:
Ok, it's just that the name of the category will show up in the CP, and it feel weird to actually in UI, and it feel weird to precise that this apply to "jupyter notebook"
![screen shot 2015-10-12 at 18 01 55](https://cloud.githubusercontent.com/assets/335567/10442839/5e28140c-710b-11e5-949c-6369eb921b3f.png). 

Feel a bit like [Truman show](https://youtu.be/Inscky6EyQ8?t=1m31s).

ellisonbg replied:
Yeah, you make a good point about how the names are described in the UI. I
did change it slightly in the UI as well to better reflect it - but still
not pretty...

On Mon, Oct 12, 2015 at 9:05 PM, Matthias Bussonnier <
notifications@github.com> wrote:

> In notebook/static/notebook/js/actions.js
> <https://github.com/jupyter/notebook/pull/598#discussion_r41817844>:
>
> >      // and uniformize/fill in missing pieces in of an action.
> >      var _prepare_handler = function(registry, subkey, source){
> > -        registry['ipython.'+subkey] = {};
> > -        registry['ipython.'+subkey].help = source[subkey].help||subkey.replace(/-/g,' ');
> > -        registry['ipython.'+subkey].help_index = source[subkey].help_index;
> > -        registry['ipython.'+subkey].icon = source[subkey].icon;
> > +        registry['jupyter-notebook:'+subkey] = {};
> > +        registry['jupyter-notebook:'+subkey].help = source[subkey].help||subkey.replace(/-/g,' ');
> > +        registry['jupyter-notebook:'+subkey].help_index = source[subkey].help_index;
> > +        registry['jupyter-notebook:'+subkey].icon = source[subkey].icon;
>
> Ok, it's just that the name of the category will show up in the CP, and it
> feel weird to actually in UI, and it feel weird to precise that this apply
> to "jupyter notebook"
> [image: screen shot 2015-10-12 at 18 01 55]
> <https://cloud.githubusercontent.com/assets/335567/10442839/5e28140c-710b-11e5-949c-6369eb921b3f.png>.
>
>
> Feel a bit like Truman show <https://youtu.be/Inscky6EyQ8?t=1m31s>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/notebook/pull/598/files#r41817844>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

sccolbert replied:
In the long term, we will augment commands with a one-line description, which can be used for presentation in the UI. We don't want to choose overly verbose command names now, just for the sake of UI presentation.

minrk replied:
It seems weird that we would namespace `jupyter-notebook` in our own Jupyter-specific command namespace when we felt `notebook` was a sufficiently unique key in the entire namespace of Python packages. Seems like either `notebook` is okay here or `notebook` wasn't an okay choice for the Python package name.

ellisonbg commented:
@Carreau as to why omit the "selected". Initially we started with that, but it felt one-step-too-verbose, when a significant fraction of the time, the user will be applying actions to the active object.

Another thing that we talked about today. For actions that make sense on multiple (marked) cells, the simple versions of commands should apply to:

* All marked cells if any are marked (multiple selection). In this case, *not* the selected cell - it acts like a cursor.
* The selected cell if none are marked.

The alternative is to have 3 versions of all actions on cells:

1. `-selected-`
2. `-marked-`
3. `-all-`

That seems overly verbose. he approach implemented in this PR automatically provides a natural way of have actions that either the selected of mulitple marked cells.

Carreau commented:
The verbosity will not be annoying for user to type once we have fuzzy-matching, but I understand your point. Though I dont' think verbosity should be scarified on action that will almost never be used through the command palette (run current cell, move cursor up or previous cell). 
These are mostly for user to do write compound actions, so are probably better explicit IMHO. 

I see you point though, but keep in mind that many user are **not** native english, and that they will be exposed to english UI, and that while english user will find some action name obvious, it will not be the case for non-english.

ellisonbg commented:
@Carreau yes, I get your point about english versus non-english users. It was actually very helpful to have @SylvainCorlay doing this work with us as he had  similar feedback as you. One suggestion was to use German for all of these as it is a quite formal language with more rigid and structured grammar rules. But I think `paste-zelle unterhalb` might be difficult to sell to our users.

ellisonbg commented:
@Carreau yes, I like your idea of being able to hide some actions from the command pallete. We are pretty sure that some of them won't ever be run that way. Pinging @sccolbert @blink1073 on that idea.

Carreau commented:
> @Carreau yes, I like your idea of being able to hide some actions from the command pallete. We are pretty sure that some of them won't ever be run that way. Pinging @sccolbert @blink1073 on that idea.

We are all consenting adults. Hide by default action stat start with `_` and show them only if user search pattern that start with  `_` ? (like private method completion in Python)

sccolbert commented:
+1 on having a prefix to distinguish advanced commands from every-day commands. I might choose `$` over `_` as the prefix though, to prevent any misunderstanding that the commands are private.

minrk commented:
+1 to most of this, I think it's a great improvement. Thanks for doing a thoughtful pass on the names. Only a couple of questions:

- inline comment about `jupyter-notebook`. Seems like `notebook` is enough, or possibly `jupyter`, but `jupyter-notebook` seems redundant at least once.
- is there any reasoning behind the switch from `.` to `:` that I missed? I don't care much, but have a slight preference for `.`.

sccolbert commented:
The choice for `:` is because that's what Phosphor is planning to use. It's also what Atom uses to denote a package namespace.

blink1073 commented:
+1 for a prefix other than `_`, I'm fine with `$`.

minrk commented:
Following existing convention is a good reason, thanks.

Carreau commented:
I'm fine with either $ or _ , the old hackpad notes for commands is [here](https://jupyter.hackpad.com/Jupyter-Commands-sDnigMnVaLr)

Carreau commented:
Seem to need rebase (sorry about that)

ellisonbg commented:
No problem, I figure I would have to rebase.

On Wed, Oct 14, 2015 at 2:13 PM, Matthias Bussonnier <
notifications@github.com> wrote:

> Assigned #598 <https://github.com/jupyter/notebook/pull/598> to @ellisonbg
> <https://github.com/ellisonbg>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/notebook/pull/598#event-435531977>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

ellisonbg commented:
Rebased

@minrk On the verbose `jupyter-notebook` namespace. A few reasons why more verbose. First, users should eventually never see this unlike the Python package which uses conda/pip install. Second, because the browser doesn't have any way to enforce global name uniqueness, we have to worry about someone, somewhere, someday calling something `notebook` other than us.

But maybe a bigger question. I would like us to start using a unique prefix approach for CSS, etc. Alot of this will start to be scoped at the level of our JS npm packages. For example, it would be great if the command and CSS namespace for the notebook component could be something like `j-notebook` or `jupyter-notebook`, etc.

What global name prefix to people like? `jp`? `jpy`?

@sccolbert @blink1073 @minrk @Carreau @jdfreder @dwillmer @jasongrout @SylvainCorlay

jdfreder commented:
I prefer `jpy-`

SylvainCorlay commented:
And a notebook file format would be `.jpynb` ?

ellisonbg commented:
OK I have rebased, and I think fixed the tests. I think we should wait to add the command hiding in a separate PR. I like the idea of using `$`, but also want to make sure we don't want to express that as a separate key in the command structure.

Remaining things to solve for this:

- [ ] Which command prefix to use?
- [ ] Any other command name changes?

minrk commented:
> First, users should eventually never see this unlike the Python package which uses conda/pip install

How is that possible, when the key is a public API and in the UI for executing commands?

> What global name prefix to people like? jp? jpy?

`jpy` seems reasonable, though I have some preference for `jp`, since I would like to avoid `py` in abbreviations as it looks like it's another `foopy` Python reference.

> And a notebook file format would be .jpynb?

Ha, no.

ellisonbg commented:
On Thu, Oct 15, 2015 at 10:57 AM, Min RK <notifications@github.com> wrote:

> First, users should eventually never see this unlike the Python package
> which uses conda/pip install
>
> How is that possible, when the key is a public API and in the UI for
> executing commands?
>
Sorry I wasn't clear. Yes, when people are writing code or keyboard
shortcuts against the API, they will see those. But in the UI, I would like
to have each command namespace also have a human readable form that
non-customizing users will see.


> What global name prefix to people like? jp? jpy?
>
> jpy seems reasonable, though I have some preference for jp, since I would
> like to avoid py in abbreviations as it looks like it's another foopy
> Python reference.
>

Yeah, I am with you on this. I think `jp` is probably best favorite so far.
I will update that in this PR unless there are other ideas.


> And a notebook file format would be .jpynb?
>
> Ha, no.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/notebook/pull/598#issuecomment-148411961>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

jdfreder commented:
jp is a-o-k with me

Carreau commented:
Still need rebase after Jon's PR was merged. 

and technically the prefix in **this** case is shown in user UI. The command palette **does** classify action in groups, and these groups are named based on the prefix. 

Unless you had yet another API to have a prefix/group name mapping.

and wether or not we prefix with `notebook` or `jupyter-notebook` nothing prevent extension to register under the same name.

 
![screen shot 2015-10-15 at 09 57 21](https://cloud.githubusercontent.com/assets/335567/10521176/2b975f24-7323-11e5-9db4-500bac3b1e4e.png)

Carreau commented on notebook/tests/notebook/markdown.js:
```diff
@@ -30,7 +30,7 @@ casper.notebook_test(function () {
         var cell = IPython.notebook.get_selected_cell();
         cell.set_text('*Baz*');
-        $("button[data-jupyter-action='ipython.run-select-next']")[0].click();
+        $("button[data-jupyter-action='ipython.run-cell-and-select-next']")[0].click();
```
`jupyter-notebook:`

jdfreder replied:
`jp-notebook:` now :)

Carreau commented:
And for the record for the **actions** I would prefer `jupyter-notebook` -> `notebook`. For the non-user visible prefixes I have no opinion on `jp` vs `jpy` vs something else.

minrk commented:
jupyter-notebook, jp, then? Works for me.

jdfreder commented:
> jupyter-notebook, jp,

+1

jdfreder commented:
@ellisonbg I added a few more comments inline and this needs a rebase.  I think we should try to get this in so it doesn't grow stale again.

ellisonbg commented:
OK I have addressed most comments. I talked more with Chris, Jason and Sylvain about the action name prefix naming.

* In the long run, if needed, we can provide a end-user-human-readable version of the command namespace for usage in UI elements.
* For this PR, I mostly want to get the programmatic one right.
* The feeling here is that the command namespaces should roughly follow JS package names. From this perspective, the notebook package will be `jupyter-js-notebook` so the command prefix would make sense as `jupyter-notebook`.
* Also feeling here like they should be verbose (more verbose than CSS) and developer-human-readable, so `jp-notebook` is too opaque.
* These will be scoped at the level of major components in the UI (what we are calling plugins on the phosphor side of things). When all is said and done, there will tons of plugins, many of which we don't develop. Because of this, I think it will be very useful to have the extra `jupyter-` in the namespace to group our components together:

`jupyter-texteditor`
`jupyter-terminal`
`jupyter-debugger`
`jupyter-filebrowser`
etc.

Based on these discussions, the current state of the PR is to use `jupyter-notebook`.

As an aside, I think it would be helpful to use a different name (maybe the shorter `jp-notebook`) for our CSS prefix to more easily allow searching through our codebase for developers.

Thoughts?

Carreau commented:
> `jupyter-js-notebook` I'm still unconvinced about `-js-` in the name.

-  Because it is/will be written in TS,
- because no other package in JS world do. 
- we don't have jupyter-py-notebook.
(- I want to punch TS in the face even more each time I have a package with - or _ in it and it fails in even weirder way than usual, but that's unrelated. ) 


The fact that the toolbar is not in env is weird, and yes `env` =! of `Jupyter` because env is not and should not be global.

ellisonbg commented:
Yeah, I am not thrilled about the `-js-` in the middle of all our JS repo names. But the only way I see around this would be for us to start a new `jupyterjs` GitHub org for all of our JS packages. Also, because everything gets compiled down to JS and people use the libraries from JS, I think it should be JS (it is pohsphorjs, not phosphorts).

Carreau commented:
Maybe fair enough for notebook, but really why not just `jupyter-input-area`, `jupyter-output-are`, `jupyter-services`, `jupyter-cell` ?

ellisonbg commented:
We could, but in the long run we are going to end up with separate JS and
Python packages for the same thing. That gets tough to avoid our own name
collisions or even name confusions. It gets tough for folks to navigate and
understand the difference between "notebook" and "jupyter-notebook" and
"jupyter-notebook-plugin"...

It might not be so bad though...

On Tue, Oct 20, 2015 at 11:42 AM, Matthias Bussonnier <
notifications@github.com> wrote:

> Maybe fair enough for notebook, but really why not just jupyter-input-area,
> jupyter-output-are, jupyter-services, jupyter-cell ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/notebook/pull/598#issuecomment-149608839>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

Carreau commented on docs/source/examples/Notebook/Custom Keyboard Shortcuts.ipynb:
```diff
@@ -114,7 +114,7 @@
     "%%javascript\n",
     "\n",
-    "Jupyter.keyboard_manager.command_shortcuts.add_shortcut('r', 'ipython.change-selected-cell-to-raw-cell');"
+    "Jupyter.keyboard_manager.command_shortcuts.add_shortcut('r', 'jp-notebook:change-cell-to-raw');"
```
jp-notebook here, you are using jupyter-notebook everywhere else.

Carreau commented:
Can you check the pager, click on `x` does not close the pager for me anymore.

Carreau commented:
And let's merge once the jp-notebook is fixed, and iterate.

ellisonbg commented:
OK, after Travis is happy, this is ready to merge. I have also figured out why the pager is broken, but it is also broken in master so I will submit another PR to fix...

Carreau commented:
> I have also figured out why the pager is broken, but it is also broken in master so I will submit another PR to fix...

Seem fair enough. 

Traavis is happy, rereading one last time and I'll press the GB.
